source/userguide/auto_annotator.rst

```diff
@@ -0,0 +1,489 @@
+If the selected English model is not in the current environment it will automatically be downloaded. Refer to Spacy's documentation to learn more about their `English models <https://spacy.io/models/en>`_.
+
+
+Creating a Custom Annotator
+---------------------------
+The :class:`SpacyAnnotator` is a subclass of the abstract base class :class:`Annotator`.
+The functionality for annotating and unannotating files is contained in :class:`Annotator` itself.
+A developer simply needs to implement two methods to create a custom annotator.
+
+
+Custom Annotator Boilerplate Code
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This section includes boilerplate code to build a :class:`CustomAnnotator` class to which you can add to your own python file, let's call it :attr:`custom_annotator.py`
+There are two "TODO"s. To implement a :class:`CustomAnnotator` class a developer has to implement the :meth:`parse` and :meth:`valid_entity_check` methods.
+
+.. code-block:: python
+
+	class CustomAnnotator(Annotator):
+		""" Custom Annotator class used to generate annotations.
+		"""
+
+		def __init__(self, app_path, config=None):
+			super().__init__(app_path=app_path, config=config)
+			
+			# Add additional attributes if needed
+
+		def parse(self, sentence, entity_types=None):
+			""" 
+			Args:
+				sentence (str): Sentence to detect entities.
+				entity_types (list): List of entity types to parse. If None, all
+					possible entity types will be parsed.
+			Returns: entities (list): List of entity dictionaries.
+			"""
+
+			# TODO: Add custom parse logic
+
+			return entities
+
+		def valid_entity_check(self, entity):
+			""" 
+			Args: entity (str): Name of entity to annotate.
+			Returns: bool: Whether entity is valid.
+			"""
+			entity = entity.lower().strip()
+			supported_entities = [
+
+			# TODO: Add list of supported entity names
+			
+			]
+			return entity in supported_entities
+	
+	if __name__ == "__main__":
+		custom_annotator = CustomAnnotator(app_path="hr_assistant")
+		custom_annotator.annotate()
+
+Entities returned by :attr:`parse()` must have the following format:
+
+.. code-block:: python
+
+	entity = { 
+		"body": (substring of sentence), 
+		"start": (start index), 
+		"end": (end index + 1), 
+		"dim": (entity type), 
+		"value": (resolved value, if it exists), 
+	}
+
+To run your custom Annotator, simply run in the command line: :attr:`python custom_annotator.py`.
+To run unannotation with your custom Annotator, change the last line in your script to :attr:`custom_annotator.unannotate()`.
+
+Getting Custom Parameters from the Config
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+:attr:`spacy_model` is an example of an optional parameter in the config that is relevant only for a specific :class:`Annotator` class.
+
+.. code-block:: python
+
+	AUTO_ANNOTATOR_CONFIG = { 
+		... 
+		"spacy_model": "en_core_web_md",
```

Comment:
I haven't yet tried using another language model but I have seen on Spacy's website that the output of NER for models that support it uses the same interface, so it should likely work so long as duckling also supports that language as well

A developer can't use a non-english model currently since an exception would be thrown if the model is not in this list: [EN_CORE_WEB_SM, EN_CORE_WEB_MD, EN_CORE_WEB_LG]

Reply:
I see - can you run a quick test to see if that can work with non-english language? in that case, we should allow support for non-english language, but just add a warning if it's not in the list.

Reply:
I was able to load up a Spacy Spanish annotator using es_core_news_lg. I had it parse the following text:

`text = "El domingo, el Gobierno de Navarra, que ha registrado 675,3 casos por cada 100.000 habitantes en las dos últimas semanas, ordenó que todos los bares y restaurantes reduzcan su capacidad al 30% y cierren a las 22.00 horas."`

It was only able to identify "Gobierno de Navarra" as location and missed the numbers and time-related phrases. When I passed the same text into duckling (language="es"), duckling is able to identify many more entities (but not able to collapse candidates to a final set of course).

Will go ahead and add the support and warning as suggested

Reply:
Done